File: Scripts/Python/Aug_27/polynomial.py
#!/usr/bin/env python
# File: polynomial.py
# Name: D.Saravanan
# Date: 27/08/2021

""" Script to construct the polynomial interpolation using Lagrange polynomial """

import numpy as np
import matplotlib
import matplotlib.pyplot as plt
matplotlib.use('Agg')

def interpolate(fnodes, xnodes, x):

    M, N = len(x), len(xnodes)
    res = np.zeros(M)

    for i in range(0, N):
        res = res + fnodes[i] * lagrange(xnodes, x, i)
    
    return res

def lagrange(xnodes, x, i):

    M, N = len(x), len(xnodes)
    val = np.ones(M)

    for j in range(0, N):

        if i == j:
            continue

        val = val * (x - xnodes[j]) / (xnodes[i] - xnodes[j])

    return val


N = 9

xnodes = np.linspace(-1,1,N)
fnodes = np.sin(np.pi*xnodes)
#fnodes = 1/(1 + 25 * xnodes**2)

x = np.linspace(-1,1,1000)
f = np.sin(np.pi*x)
#f = 1/(1 + 25 * x**2)

fintp = interpolate(fnodes, xnodes, x)

error = max(abs(f - fintp)/abs(f))
print(error)

fig, ax = plt.subplots()
ax.plot(x, f, 'r')
ax.plot(x, fintp, 'g')
ax.plot(xnodes, fnodes, 'b*')
ax.set(xlim=(-1, 1), ylim=(-1.2, 1.2))
ax.set_title(r"Lagrange polynomial interpolation", fontsize=11)
ax.grid(True)
plt.savefig('graph.png')

File: Scripts/Julia/Aug_27/polynomial.jl
#!/usr/bin/env julia
# File: polynomial.jl
# Name: D.Saravanan
# Date: 27/08/2021

""" Program to construct the polynomial interpolation
using Lagrange polynomial """

import PyPlot
const plt = PyPlot
plt.rc("pgf", texsystem="pdflatex")
plt.rc("font", family="serif", weight="normal", size=8)
plt.rc("axes", labelsize=10, titlesize=10)
plt.rc("figure", titlesize=10)
plt.rc("text", usetex="True")
using LaTeXStrings

function interpolate(fnodes, xnodes, x)
    
    M, N = length(x), length(xnodes)
    res = zeros(M)

    for i in 1:N
        res = res + fnodes[i] * lagrange(xnodes, x, i)
    end

    return res
end

function lagrange(xnodes, x, i)
    
    M, N = length(x), length(xnodes)
    val = ones(M)

    for j in 1:N

        if i == j
            continue
        end

        val = val .* (x .- xnodes[j]) / (xnodes[i] - xnodes[j])
    end

    return val
end


N = 9

xnodes = LinRange(-1,1,N)
fnodes = sin.(pi .* xnodes)

# true function calculated with equi-spaced nodes
x = LinRange(-1,1,1000)
f = sin.(pi .* x)

fintp = interpolate(fnodes, xnodes, x)

error = maximum(abs.(f .- fintp) ./abs.(f))
println(error)

fig, ax = plt.subplots()
ax.plot(x, f, c="olive", ls="dotted", label=L"$\sin(\pi x)$")
ax.plot(x, fintp, c="teal", label=L"$Interpolation$")
ax.plot(xnodes, fnodes, "m.", label=L"$N = 9$")
ax.set(xlim=(-1, 1), ylim=(-1.1, 1.1))
ax.set(xlabel=L"$x$", ylabel=L"$\sin(\pi x)$")
ax.set_title(raw"Lagrange polynomial interpolation")
ax.grid(true); ax.legend()
plt.savefig("polynomial.pdf")
